lib/newProfile.ml

```diff
@@ -0,0 +1,200 @@
+          if String.equal name name' then Some (r,cnt+cnt')
+          else Some (l +. r, cnt+cnt'))
+          sum next
+      in
+      sums := (start', next) :: rest;
+      sum, dur
+
+let leave ?time name ?(args=[]) ?last () =
+  let time = gettimeopt time in
+  let sum, dur = leave_sums ~time name () in
+  let sum = List.map (fun (name, (t, cnt)) ->
+      name, MiniJson.String
+        (Printf.sprintf "%.3G us, %d %s" (t *. 1E6) cnt (CString.plural cnt "call")))
+      (CString.Map.bindings sum)
+  in
+  let args = ("subtimes", MiniJson.Record sum) :: args in
+  duration ~time name "E" ~args ?last ()
+
+let make_mem_diff ~(mstart:Gc.stat) ~(mend:Gc.stat) =
+  (* XXX we could use memprof to get sampling stats instead?
+     eg stats about how much of the allocation was collected
+     in the same span vs how much survived it *)
+  let major = mend.major_words -. mstart.major_words in
+  let minor = mend.minor_words -. mstart.minor_words in
+  let pp tdiff = MiniJson.String (Printf.sprintf "%.3G w" tdiff) in
+  [("major",pp major); ("minor", pp minor)]
+
+(* NB: "process" and "init" are unconditional because they don't go
+   through [profile] and I'm too lazy to make them conditional *)
+let components =
+  match Sys.getenv_opt "COQ_PROFILE_COMPONENTS" with
+  | None -> CString.Pred.full
+  | Some s ->
+    List.fold_left (fun cs c -> CString.Pred.add c cs)
+      CString.Pred.empty
+      (String.split_on_char ',' s)
+
+let profile name ?args f () =
+  if not (is_profiling ()) then f ()
+  else if CString.Pred.mem name components then begin
+    let args = Option.map (fun f -> f()) args in
+    enter name ?args ();
+    let mstart = Gc.quick_stat () in
+    let v = try f ()
+      with e ->
+        let e = Exninfo.capture e in
+        let mend = Gc.quick_stat () in
+        let args = make_mem_diff ~mstart ~mend in
+        leave name ~args ();
+        Exninfo.iraise e
+    in
+    let mend = Gc.quick_stat () in
+    let args = make_mem_diff ~mstart ~mend in
+    leave name ~args ();
+    v
+  end
+  else begin
+    enter_sums ();
+    let v = try f ()
+      with e ->
+        let e = Exninfo.capture e in
+        ignore (leave_sums name () : _ * _);
+        Exninfo.iraise e
+    in
+    ignore (leave_sums name () : _ * _);
+    v
+  end
+
+let init opts =
+  let () = assert (not (is_profiling())) in
+  match opts with
+  | None ->
+    accu := No
+  | Some { file } ->
+    let ch = open_out file in
+    accu := File ch;
+    f "{ \"traceEvents\": [\n";
+    enter ~time:global_start "process" ();
+    enter ~time:global_start "init" ();
+    leave "init" ~args:(make_mem_diff ~mstart:global_start_stat ~mend:(Gc.quick_stat())) ()
+
+let finish () = match !accu with
+  | No -> ()
+  | File ch ->
+    leave "process"
+      ~last:""
+      ~args:(make_mem_diff ~mstart:global_start_stat ~mend:(Gc.quick_stat()))
+      ();
+    Printf.fprintf ch "],\n\"displayTimeUnit\": \"us\" }";
+    close_out ch
+
+let () = at_exit finish
```

Comment:
Please make this explicit, otherwise this can't work with `coq-lsp` and other systems that handle multilpe documents.